GMP 6.3.0 fails to build due to GLIBCXX version incompatibility with self-compiled gcc

### Steps To Reproduce

Seeing this on the buildbot when building our own gcc
```
[gmp-6.3.0] [spkg-install] checking build system compiler gcc... yes
[gmp-6.3.0] [spkg-install] checking for build system preprocessor... gcc -E
[gmp-6.3.0] [spkg-install] checking for build system executable suffix... 
[gmp-6.3.0] [spkg-install] checking whether build system compiler is ANSI... yes
[gmp-6.3.0] [spkg-install] checking for build system compiler math library... -lm
[gmp-6.3.0] [spkg-install] checking whether we are using the GNU C++ compiler... yes
[gmp-6.3.0] [spkg-install] checking whether g++ accepts -g... yes
[gmp-6.3.0] [spkg-install] checking C++ compiler g++  -O2 -pedantic -fomit-frame-pointer -m64 -mtune=znver2 -march=znver2... no, std iostream, program does not run
[gmp-6.3.0] [spkg-install] checking C++ compiler g++  -g -O2... no, std iostream, program does not run
[gmp-6.3.0] [spkg-install] configure: error: C++ compiler not available, see config.log for details
[gmp-6.3.0] [spkg-install] Error configuring GMP (with CFLAGS unset).
[gmp-6.3.0] [spkg-install] Consult /var/lib/buildbot/worker/sage_git/build/local/var/tmp/sage/build/gmp-6.3.0/src/config.log for for details.
[gmp-6.3.0] ************************************************************************
[gmp-6.3.0] Error installing package gmp-6.3.0
[gmp-6.3.0] ************************************************************************
[gmp-6.3.0] Please email sage-devel (http://groups.google.com/group/sage-devel)
[gmp-6.3.0] explaining the problem and including the log files
[gmp-6.3.0]   /var/lib/buildbot/worker/sage_git/build/logs/pkgs/gmp-6.3.0.log
[gmp-6.3.0] and
[gmp-6.3.0]   /var/lib/buildbot/worker/sage_git/build/config.log
[gmp-6.3.0] Describe your computer, operating system, etc.
[gmp-6.3.0] If you want to try to fix the problem yourself, *don't* just cd to
[gmp-6.3.0] /var/lib/buildbot/worker/sage_git/build/local/var/tmp/sage/build/gmp-6.3.0 and type 'make' or whatever is appropriate.
[gmp-6.3.0] Instead, the following commands setup all environment variables
[gmp-6.3.0] correctly and load a subshell for you to debug the error:
[gmp-6.3.0]   (cd '/var/lib/buildbot/worker/sage_git/build/local/var/tmp/sage/build/gmp-6.3.0' && '/var/lib/buildbot/worker/sage_git/build/sage' --buildsh)
[gmp-6.3.0] When you are done debugging, you can type "exit" to leave the subshell.
[gmp-6.3.0] ************************************************************************
```
Actual error from config.log:
```
Test compile: std iostream
configure:10620: g++  -g -O2 conftest.cc >&5
configure:10623: $? = 0
configure:10627: ./a.out || ./b.out || ./a.exe || ./a_out.exe || ./conftest
./a.out: /usr/lib/x86_64-linux-gnu/libstdc++.so.6: version `GLIBCXX_3.4.32' not found (required by ./a.out)
./configure: line 10628: ./b.out: No such file or directory
./configure: line 10628: ./a.exe: No such file or directory
./configure: line 10628: ./a_out.exe: No such file or directory
./configure: line 10628: ./conftest: No such file or directory
configure:10630: $? = 127
```


### Expected Behavior

The GMP configure test should link against our own libstdc++ instead of the (too old) system libstdc++

### Actual Behavior

conftest tries os libstdc++ and fails

### Additional Information

_No response_

### Environment

```markdown
- **OS**: Debian 11 64-bit
- **Sage Version**: 10.4.beta7
```


### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

--- a/build/pkgs/gmp/spkg-install.in
+++ b/build/pkgs/gmp/spkg-install.in
@@ -98,10 +98,16 @@ if [ "$SAGE_FAT_BINARY" = "yes" ]; then
     esac
 fi
 
+# Make configure tests use our own libstdc++ if it is available
+# See https://github.com/sagemath/sage/issues/38085
+CXXFLAGS_CONFIGURE="-Wl,-rpath -Wl,$SAGE_LOCAL/lib"
 
 # Pre-configure GMP to get the settings it would use if CFLAGS were empty:
 echo "Checking what CFLAGS GMP would use if they were empty..."
-(unset CFLAGS CPPFLAGS CXXFLAGS && ./configure $GMP_CONFIGURE) &>configure-empty.log
+(
+    unset CFLAGS CPPFLAGS CXXFLAGS &&
+    CXXFLAGS=$CXXFLAGS_CONFIGURE ./configure $GMP_CONFIGURE
+) &>configure-empty.log
 if [ $? -ne 0 ]; then
     # Output the log of the failed configure run
     cat configure-empty.log
@@ -163,7 +169,7 @@ echo "You can set GMP_CONFIGURE to pass additional parameters."
 # Clear the cache of the previous configure run
 find . -name config.cache -exec rm -f {} \;
 
-./configure --prefix="$SAGE_LOCAL" --libdir="$SAGE_LOCAL/lib" $GMP_CONFIGURE
+CXXFLAGS=$CXXFLAGS_CONFIGURE ./configure --prefix="$SAGE_LOCAL" --libdir="$SAGE_LOCAL/lib" $GMP_CONFIGURE
 if [ $? -ne 0 ]; then
     echo >&2 "Error configuring GMP. (See above for the options passed to it.)"
     exit 1

